rust/ip-packet/src/make.rs

```diff
@@ -38,58 +47,75 @@ pub fn icmp_response_packet(packet: IpPacket<'static>) -> MutableIpPacket<'stati
     )
 }
 
-enum IcmpKind {
+#[cfg_attr(test, derive(Debug, test_strategy::Arbitrary))]
+pub(crate) enum IcmpKind {
     Request,
     Response,
 }
 
-fn icmp_packet(
-    src: IpAddr,
-    dst: IpAddr,
+pub(crate) fn icmp4_packet_with_options(
+    src: Ipv4Addr,
+    dst: Ipv4Addr,
     seq: u16,
     identifier: u16,
     kind: IcmpKind,
+    ip_header_length: u8,
 ) -> MutableIpPacket<'static> {
-    match (src, dst) {
-        (IpAddr::V4(src), IpAddr::V4(dst)) => {
-            use crate::{
-                icmp::{
-                    echo_request::{IcmpCodes, MutableEchoRequestPacket},
-                    IcmpTypes, MutableIcmpPacket,
-                },
-                ip::IpNextHeaderProtocols,
-                MutablePacket as _, Packet as _,
-            };
-
-            let mut buf = vec![0u8; 60];
-
-            ipv4_header(src, dst, IpNextHeaderProtocols::Icmp, &mut buf[..]);
-
-            let mut icmp_packet = MutableIcmpPacket::new(&mut buf[20..]).unwrap();
+    use crate::{
+        icmp::{
+            echo_request::{IcmpCodes, MutableEchoRequestPacket},
+            IcmpTypes, MutableIcmpPacket,
+        },
+        ip::IpNextHeaderProtocols,
+        MutablePacket as _,
+    };
+
+    let ip_header_bytes = ip_header_length * 4;
+    let mut buf = vec![0u8; 60 + ip_header_bytes as usize];
+
+    ipv4_header(
+        src,
+        dst,
+        IpNextHeaderProtocols::Icmp,
+        ip_header_length,
+        &mut buf[20..],
+    );
+
+    let mut icmp_packet =
+        MutableIcmpPacket::new(&mut buf[(20 + ip_header_bytes as usize)..]).unwrap();
+
+    match kind {
+        IcmpKind::Request => {
+            icmp_packet.set_icmp_type(IcmpTypes::EchoRequest);
+            icmp_packet.set_icmp_code(IcmpCodes::NoCode);
+        }
+        IcmpKind::Response => {
+            icmp_packet.set_icmp_type(IcmpTypes::EchoReply);
+            icmp_packet.set_icmp_code(IcmpCodes::NoCode);
+        }
+    }
 
-            match kind {
-                IcmpKind::Request => {
-                    icmp_packet.set_icmp_type(IcmpTypes::EchoRequest);
-                    icmp_packet.set_icmp_code(IcmpCodes::NoCode);
-                }
-                IcmpKind::Response => {
-                    icmp_packet.set_icmp_type(IcmpTypes::EchoReply);
-                    icmp_packet.set_icmp_code(IcmpCodes::NoCode);
-                }
-            }
+    icmp_packet.set_checksum(0);
 
-            icmp_packet.set_checksum(0);
+    let mut echo_request_packet = MutableEchoRequestPacket::new(icmp_packet.packet_mut()).unwrap();
+    echo_request_packet.set_sequence_number(seq);
+    echo_request_packet.set_identifier(identifier);
```

Comment:
Shouldn't this also be conditional on the kind?

Reply:
it's like this on main https://github.com/firezone/firezone/blob/main/rust/ip-packet/src/make.rs#L121-L122 the reason why this works is because the parsing is the same for echo request and reply, I think it's okay to leave it like here.